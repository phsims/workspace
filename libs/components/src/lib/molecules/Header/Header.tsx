'use client';
import { ReactNode, useState } from 'react';
import { useRouter } from 'next/navigation';
import { Box } from '@mui/material';

import {
  AppBar,
  Toolbar,
  IconButton,
  Typography,
  Drawer,
  ListItemButton,
  MenuList,
  Container,
  Button,
} from '@mui/material';
import MenuRoundedIcon from '@mui/icons-material/MenuRounded';

interface NavProps {
  href: string;
  label: string;
  icon?: ReactNode;
}

export interface HeaderProps {
  logo?: ReactNode;
  headerNav: NavProps[];
  title: string;
  titleColor?: string;
  auth?: ReactNode;
}

export function Header({
  headerNav,
  title,
  titleColor = 'primary',
  logo = null,
  auth = null,
}: HeaderProps) {
  const router = useRouter();
  const [open, setOpen] = useState(false);

  const toggleDrawer = (newOpen: boolean) => () => {
    setOpen(newOpen);
  };

  const handleNavClick = (href: string) => {
    router.push(href);
    setOpen(false);
  };

  return (
    <>
      <AppBar
        position="static"
        color="transparent"
        sx={{ p: ' 1rem 0', flexGrow: 1 }}
      >
        <Container>
          <Toolbar>
            {headerNav && (
              <IconButton
                edge="start"
                color="inherit"
                aria-label="menu"
                sx={{ mr: 2 }}
                onClick={toggleDrawer(true)}
              >
                <MenuRoundedIcon sx={{ height: '28px', width: '28px' }} />
              </IconButton>
            )}

            <Button
              onClick={() => {
                router.push('/');
              }}
              sx={{
                p: 0,
                ':hover': {
                  bgcolor: 'transparent',
                },
              }}
            >
              {logo && <Box sx={{ mr: 2 }}>{logo}</Box>}

              <Typography
                variant="h2"
                component="div"
                sx={{
                  flexGrow: 1,
                  color: titleColor,
                  textTransform: 'capitalize',
                }}
              >
                {title}
              </Typography>
            </Button>
            {auth}
          </Toolbar>
        </Container>
      </AppBar>
      {headerNav && (
        <Drawer open={open} onClose={toggleDrawer(false)}>
          <MenuList dense sx={{ width: 320, p: 4 }}>
            {headerNav.map(({ href, label, icon }) => (
              <ListItemButton
                key={label}
                onClick={() => {
                  handleNavClick(href);
                }}
                sx={{
                  display: 'flex',
                  justifyContent: 'flex-start',
                  alignItems: 'center',
                }}
              >
                {icon && <Box sx={{ mr: 2 }}>{icon}</Box>}
                {label}
              </ListItemButton>
            ))}
          </MenuList>
        </Drawer>
      )}
    </>
  );
}

export default Header;
