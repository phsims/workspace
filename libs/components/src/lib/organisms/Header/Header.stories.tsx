import type { Meta, StoryObj } from '@storybook/react';
import { Header } from './Header';

import { within } from '@storybook/testing-library';
import { expect } from '@storybook/jest';

const meta: Meta<typeof Header> = {
  component: Header,
  title: '@organisms/Header',
};
export default meta;
type Story = StoryObj<typeof Header>;

export const Primary = {
  args: {
    title:'My cool app',
    headerNav:[
      { link: '/link1', text: 'link 1' },
      { link: '/link2', text: 'link 2' },
    ]
  },
};

export const Heading: Story = {
  args: {...Primary.args},
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    expect(canvas.getByText(/My cool app /gi)).toBeInTheDocument();
  },
};
