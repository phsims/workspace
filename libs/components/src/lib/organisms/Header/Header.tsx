'use client';
import { usePathname } from 'next/navigation';
import router from 'next/router';
import {
  Box,
  AppBar,
  Toolbar,
  IconButton,
  Typography,
  Drawer,
  ListItemButton,
  MenuList,
  Container,
} from '@mui/material';
import MenuRoundedIcon from '@mui/icons-material/MenuRounded';
import { useState } from 'react';

interface NavProps {
  href: string;
  label: string;
}

export interface HeaderProps {
  headerNav: NavProps[];
  title: string;
  titleColor?: string;
}

export function Header({
  headerNav,
  title,
  titleColor = 'primary',
}: HeaderProps) {
  const pathname = usePathname();
  const [open, setOpen] = useState(false);

  const toggleDrawer = (newOpen: boolean) => () => {
    setOpen(newOpen);
  };

  return (
    <>
      <AppBar
        position="static"
        color="transparent"
        sx={{ padding: 0, flexGrow: 1 }}
      >
        <Container>
          <Toolbar>
            {headerNav && (
              <IconButton
                edge="start"
                color="inherit"
                aria-label="menu"
                sx={{ mr: 2 }}
                onClick={toggleDrawer(true)}
              >
                <MenuRoundedIcon sx={{ height: '28px', width: '28px' }} />
              </IconButton>
            )}
            <Typography
              variant="h2"
              component="div"
              sx={{ flexGrow: 1, color: titleColor, fontSize: '2rem' }}
            >
              {title}
            </Typography>
          </Toolbar>
        </Container>
      </AppBar>
      {headerNav && (
        <Drawer open={open} onClose={toggleDrawer(false)}>
          <MenuList dense sx={{ width: 320 }}>
            {headerNav.map(({ href, label }) => (
              <ListItemButton key={label} onClick={() => router.push(href)}>
                {label}
              </ListItemButton>
            ))}
          </MenuList>
        </Drawer>
      )}
    </>
  );
}

export default Header;
