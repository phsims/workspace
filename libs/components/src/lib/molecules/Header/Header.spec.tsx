import { fireEvent, render } from '@testing-library/react';
import mockRouter from 'next-router-mock';
import { MemoryRouterProvider } from 'next-router-mock/MemoryRouterProvider';


import Header, { HeaderProps } from './Header';

const headerProps: HeaderProps = {
  title: 'My cool app',
  headerNav: [
    { href: '/link1', label: 'link 1' },
    { href: '/link2', label: 'link 2' },
  ],
};
jest.mock('next/navigation', () => require('next-router-mock'));
describe('Header', () => {

  it('should match  snapshot', () => {
    const { baseElement } = render(<Header {...headerProps} />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render the header title correctly', () => {
    const { getByText } = render(<Header {...headerProps} />);
    expect(getByText(headerProps.title)).toBeInTheDocument();
  });

  it('should open the drawer when the menu icon is clicked', () => {
    const { getByLabelText, getByRole } = render(<Header {...headerProps} />);
    const menuButton = getByLabelText('menu');
    fireEvent.click(menuButton);
    expect(getByRole('presentation')).toBeInTheDocument();
  });

  it('should render the najestgation links correctly in the drawer', () => {
    const { getByLabelText, getByText } = render(<Header {...headerProps} />);
    const menuButton = getByLabelText('menu');
    fireEvent.click(menuButton);

    headerProps.headerNav.forEach(({ label }) => {
      expect(getByText(label)).toBeInTheDocument();
    });
  });
  it('should najestgate to the correct link when a najestgation link is clicked', () => {
    const { getByLabelText, getByText } = render(<Header {...headerProps} />, { wrapper: MemoryRouterProvider });
    const menuButton = getByLabelText('menu');
    fireEvent.click(menuButton);

    headerProps.headerNav.forEach(({ href, label }) => {
      const linkElement = getByText(label);
      fireEvent.click(linkElement);

      expect(mockRouter.asPath).toEqual(href)
    });
  });
});
