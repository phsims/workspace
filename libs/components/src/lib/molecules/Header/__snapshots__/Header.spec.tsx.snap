// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should match  snapshot 1`] = `
<body>
  <div>
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorTransparent MuiAppBar-positionStatic css-iu7ypf-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiContainer-root MuiContainer-maxWidthLg css-1oqqzyl-MuiContainer-root"
      >
        <div
          class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-1yo9slo-MuiToolbar-root"
        >
          <div
            class="MuiBox-root css-hawwd0"
          >
            <button
              aria-label="menu"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-edgeStart MuiIconButton-sizeMedium css-134qg7o-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-1oazvwu-MuiSvgIcon-root"
                data-testid="MenuRoundedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M4 18h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1m0-5h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1s.45 1 1 1M3 7c0 .55.45 1 1 1h16c.55 0 1-.45 1-1s-.45-1-1-1H4c-.55 0-1 .45-1 1"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-10frapl-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            <div
              class="MuiTypography-root MuiTypography-h2 css-gi57be-MuiTypography-root"
            >
              My cool app
            </div>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
          <div
            class="MuiBox-root css-1j9bsaa"
          />
        </div>
      </div>
    </header>
  </div>
</body>
`;
